src/v/cloud_topics/level_zero/stm/ctp_stm.cc

```diff
@@ -11,84 +11,141 @@
 #include "cloud_topics/level_zero/stm/ctp_stm.h"
 
 #include "bytes/iobuf.h"
-#include "bytes/iobuf_parser.h"
 #include "cloud_topics/level_zero/stm/ctp_stm_commands.h"
 #include "cloud_topics/level_zero/stm/ctp_stm_state.h"
+#include "cloud_topics/level_zero/stm/placeholder.h"
 #include "cloud_topics/types.h"
+#include "raft/consensus.h"
 #include "serde/rw/map.h"
 #include "serde/rw/uuid.h"
 #include "serde/rw/vector.h"
+#include "storage/offset_translator_state.h"
 
 #include <stdexcept>
 
 namespace experimental::cloud_topics {
 
+static cluster_epoch extract_epoch(model::record_batch&& batch) {
+    vassert(
+      batch.header().type == model::record_batch_type::dl_placeholder,
+      "Expected batch type to be dl_placeholder, got {}",
+      batch.header().type);
+    iobuf value;
+    batch.for_each_record([&value](model::record&& r) {
+        value = std::move(r).release_value();
+        return ss::stop_iteration::yes;
+    });
+
+    auto placeholder = serde::from_iobuf<dl_placeholder>(std::move(value));
+    return placeholder.id.epoch;
+}
+
+ss::future<ss::stop_iteration>
+ctp_stm_consumer::operator()(model::record_batch batch) {
+    _first_epoch = extract_epoch(std::move(batch));
+    co_return _first_epoch.has_value() ? ss::stop_iteration::yes
+                                       : ss::stop_iteration::no;
+}
+
+std::optional<cluster_epoch> ctp_stm_consumer::end_of_stream() {
+    return _first_epoch;
+}
+
 ctp_stm::ctp_stm(ss::logger& logger, raft::consensus* raft)
   : raft::persisted_stm<>(name, logger, raft) {}
 
-ss::future<> ctp_stm::do_apply(const model::record_batch& batch) {
-    _state.get_offsets().advance_insync_offset(batch.base_offset());
-    if (batch.header().type != model::record_batch_type::ctp_stm_command) {
-        co_return;
+const model::ntp& ctp_stm::ntp() const noexcept { return _raft->ntp(); }
+
+ss::future<std::optional<cluster_epoch>> ctp_stm::get_min_epoch() {
+    // Consume the first epoch from the partition starting from
+    // start offset if nothing was reconciled yet or from the last
+    // reconciled offset + 1 otherwise.
+    auto so = _raft->start_offset();
+    auto co = _raft->committed_offset();
+    auto lro = _state.get_last_reconciled_log_offset().value_or(
+      model::prev_offset(so));
```

Comment:
in header can you precisely define what min epoch is? from what i can tell it is exactly equal to the epoch of the first placeholder at or after the LRO. but when it comes to GC, such a value fitting that definition doesn't seem to be involved in the GC process defined in the RFC (at least not directly--maybe its indirectly related?). is it supposed to be equivalent to "get_inactive_epoch" in the RFC?

Reply:
> roughly equivalent

i dont' think that roughly is good enough. they don't also don't have to be actually equivalent. we need to understand that it is correct in the larger picture of the GC algorithm.

Reply:
The RFC doesn't provide a definition of the `inactive_epoch`. Here is my definition: "inactive epoch of the CTP is the epoch which is no longer referenced by the CTP". By this definition the `max_epoch -1` is an `inactive_epoch`.  Also, more than one `inactive_epoch` values could exists simultaneously for the CTP. The value defined as `min_epoch - 1` is just a `max(∀ epoch ∈ inactive_epoch)`